File: platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/org/maplibre/android/maps/IconManagerResolver.java
package org.maplibre.android.maps;

import org.maplibre.android.annotations.Icon;

import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.Map;

import timber.log.Timber;

public class IconManagerResolver {

  private IconManager iconManager;

  public IconManagerResolver(MapLibreMap maplibreMap) {
    try {
      Field annotationManagerField = MapLibreMap.class.getDeclaredField("annotationManager");
      annotationManagerField.setAccessible(true);
      AnnotationManager annotationManager = (AnnotationManager) annotationManagerField.get(maplibreMap);

      Field iconManagerField = AnnotationManager.class.getDeclaredField("iconManager");
      iconManagerField.setAccessible(true);
      iconManager = (IconManager) iconManagerField.get(annotationManager);
    } catch (Exception exception) {
      Timber.e(exception, "Could not create IconManagerResolver, unable to reflect.");
    }
  }

  @SuppressWarnings("unchecked")
  public Map<Icon, Integer> getIconMap() {
    try {
      Field field = IconManager.class.getDeclaredField("iconMap");
      field.setAccessible(true);
      return (Map<Icon, Integer>) field.get(iconManager);
    } catch (NoSuchFieldException exception) {
      Timber.e(exception, "Could not getIconMap, unable to reflect.");
    } catch (IllegalAccessException exception) {
      Timber.e(exception, "Could not getIconMap, unable to reflect.");
    }
    return new HashMap<>();
  }
}

File: platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/org/maplibre/android/maps/IconManagerResolver.kt
package org.maplibre.android.maps

import org.maplibre.android.annotations.Icon
import timber.log.Timber

class IconManagerResolver(maplibreMap: MapLibreMap?) {
    private var iconManager: IconManager? = null

    init {
        try {
            val annotationManagerField = MapLibreMap::class.java.getDeclaredField("annotationManager")
            annotationManagerField.isAccessible = true
            val annotationManager = annotationManagerField[maplibreMap] as AnnotationManager
            val iconManagerField = AnnotationManager::class.java.getDeclaredField("iconManager")
            iconManagerField.isAccessible = true
            iconManager = iconManagerField[annotationManager] as IconManager
        } catch (exception: Exception) {
            Timber.e(exception, "Could not create IconManagerResolver, unable to reflect.")
        }
    }

    val iconMap: Map<Icon, Int>
        get() {
            try {
                val field = IconManager::class.java.getDeclaredField("iconMap")
                field.isAccessible = true
                return field[iconManager] as Map<Icon, Int>
            } catch (exception: NoSuchFieldException) {
                Timber.e(exception, "Could not getIconMap, unable to reflect.")
            } catch (exception: IllegalAccessException) {
                Timber.e(exception, "Could not getIconMap, unable to reflect.")
            }
            return HashMap()
        }
}
